File: Greedy/scheduling_problem.py
"""
    The interval schedulling problem
    A computation problem that requires optimality and completeness
    
    Approach: A Greedy Algorithm

    Problem statement:
        There is only one Resouerce available of usage by number of request
        Each request utilizes the for a given period of time (r.start and r.end)
        When a request accuqiures the resours of a time t, there shouldnt be any other request that has it

        We are looking to get a solution that selects from a set R of request a subset A 
        such that #requests in A is as close as posible to #requests in original R
        considering R contains requests that dont overlap with in each other
        
    Formalizing the  problem:
        condsire R -> Set implemented as a  HasMap for each entry having a request in the form
        {int Request_id: Request(int start, int end)} 
        Search from R to produce A such that nR ~= nA fulfilling the invariant
        Invariant of the problem : r in A | r(i).end <= r(i+1).start
    
    Solution:
        function Schedule(Request R) returns (Request A):
            L <- sort(r.valuesOfEnd)
            A <- null
            loop:
                for i in L
                    for j=i+1 in L
                    if j.valueOfStart >= i.valueOfEnd then append i in A, i++
                    else 
                        i+j                        
"""

class Request:
    def __init__(self, request_id, start, end):
        self.request_id = request_id
        self.start = start
        self.end = end

def Schedule(R):
    # Sort requests by their end times
    L = sorted(R.values(), key=lambda r: r.end)
    A = []  # List to hold selected requests
    last_end_time = 0  # Track the end time of the last added request

    for request in L:
        if request.start >= last_end_time:  # If the request does not overlap
            A.append(request)  # Add to the selected requests
            last_end_time = request.end  # Update the end time

    return A  # Return the list of selected requests

File: Greedy/SchedulingProblem.java
import java.util.*;

public class SchedulingProblem {
    // Request class to hold start and end times
    static class Request {
        int id;
        int start;
        int end;

        Request(int id, int start, int end) {
            this.id = id;
            this.start = start;
            this.end = end;
        }
    }

    // Method to schedule requests
    public static List<Request> schedule(List<Request> requests) {
        // Sort requests by their end times
        requests.sort(Comparator.comparingInt(r -> r.end));
        
        List<Request> selectedRequests = new ArrayList<>();
        Request lastSelected = null;

        for (Request request : requests) {
            // If this request starts after or when the last selected request ends
            if (lastSelected == null || request.start >= lastSelected.end) {
                selectedRequests.add(request);
                lastSelected = request; // Update the last selected request
            }
        }
        
        return selectedRequests;
    }

    // Main method for testing
    public static void main(String[] args) {
        List<Request> requests = Arrays.asList(
            new Request(1, 0, 3),
            new Request(2, 1, 4),
            new Request(3, 3, 5),
            new Request(4, 5, 6),
            new Request(5, 2, 7),
            new Request(6, 6, 8),
            new Request(7, 7, 9),
            new Request(8, 8, 10),
            new Request(9, 9, 11),
            new Request(10, 10, 12)
        );

        List<Request> scheduledRequests = schedule(requests);
        for (Request req : scheduledRequests) {
            System.out.println("Request ID: " + req.id + " Start: " + req.start + " End: " + req.end);
        }
    }
}
